### DISCUSSION TITLE
Why ACE has not provided the RAW socket implemention

### DISCUSSION THREAD

**Discussion by smithAchang** [DOCGroup/ACE_TAO#2061] [Ideas]
Although using epoll mechanism， I can create many udp sockets.

But for condsidering the file handle limit by OS and some config related reason,  raw + udp protocol socket can provide some advantages,.

Why ACE has not provided the RAW socket implemention ? 

Can ACE support the feature in the future? 

**thx :)**

> ACE_ICMP_Socket is similiar but can not be used directly

**Comment by jwillemsen** [comment]
It seems nobody has created an ACE abstraction for raw sockets, feel free to create one and contribute it back to ACE (including tests)

**Reply by smithAchang** [reply]
I find there is a easy way to implement the raw socket class by reusing some codes from the `ACE_ICMP_Socket` class.
Maybe we also can  redesign the the `ACE_ICMP_Socket` class to make `ACE_RAW_Socket`  as his parent class

Can  ACE maintaining team implements the work ?  thx :)

**Comment by shuston** [comment]
ACE improvements are driven by users' need and contributions of either funding or the actual code. If you develop the ACE_RAW_Socket changes and submit a pull request it will be considered. Please make sure that you follow the ACE coding guidelines and create a set of tests for it in the ACE_wrappers/tests directory.

**Reply by smithAchang** [reply]
OK 
I will provide the work in the future , but maybe need a long delay

By the way , where can I find the ACE coding guidelines ? I find no related item in Wiki

**Reply by smithAchang** [reply]
how can I build the ACE code based the source code from github ?

---

must I download the tar from [officesite](https://download.dre.vanderbilt.edu/) to setup the compiling dependencies ?

**Comment by smithAchang** [comment]
I have got it !!!

---

[Building and Installing ACE](https://www.dre.vanderbilt.edu/~schmidt/DOC_ROOT/ACE/ACE-INSTALL.html)

git clone git@github.com:DOCGroup/ACE_TAO.git

export ACE_ROOT=/path/to/ACE_wrappers
export LD_LIBRARY_PATH=$ACE_ROOT/lib:$LD_LIBRARY_PATH



echo "ace/config-linux.h" > $ACE_ROOT/ace/config.h
echo "include $(ACE_ROOT)/include/makeinclude/platform_linux.GNU" > $ACE_ROOT/include/makeinclude/platform_macros.GNU

cd $ACE_ROOT

if [ ! -d MPC ]; then
 git clone https://github.com/DOCGroup/MPC.git MPC
fi


if [ add some new file ]; then
  modify the $ACE_ROOT/ace/ace.mpc
fi

$ACE_ROOT/bin/mwc.pl -type gnuace ACE.mwc

cd $ACE_ROOT

make
